File: day5/bf.lua
local data_tape = {
    head = 0,
    size = 256,
    tape = {}
}

function main()
    -- set up the initial section of the tape
    for i = 0, data_tape.size do
        data_tape.tape[i] = 0
    end

    if #arg > 0 then
        -- get a file and just run interpreter on it
        local file = assert(io.open(arg[1], "r"))
        local src = file:read("*all")
        file:close()

        interpreter(src)
    else
        -- otherwise fall through to repl
        repl()
    end
end

function check_brace_balance(src)
    local opens = 0
    local closes = 0

    local head = 0
    while head <= #src do
        if src:sub(head, head) == '[' then
            opens = opens + 1
        elseif src:sub(head, head) == ']' then
            closes = closes + 1
        end

        if closes > opens then
            print("[ERROR]: Unbalanced ']'.")
            os.exit(false, true)
        end

        head = head + 1
    end

    return opens - closes
end

function interpreter(src)
    -- make sure that [] are balanced, this has always technically been optional, but it's not hard or remotely complicated so I figure why not just keep doing it
    local balance = check_brace_balance(src)
    if balance > 0 then
        print("[ERROR]: Unbalanced '['.")
        os.exit(false, true)
    elseif balance < 0 then
        print("[ERROR]: Unbalanced ']'.")
        os.exit(false, true)
    end

    -- idk what the technical term for this would be but it's effectively a "goto" dispatch table
    local instructions = {
        ['>'] = function()
            data_tape.head = data_tape.head + 1
            if data_tape.head > data_tape.size then
                data_tape.size = data_tape.size * 2
                for i = data_tape.head, data_tape.size do
                    data_tape.tape[i] = 0
                end
            end
        end,
        ['<'] = function() 
            if data_tape.head > 0 then 
                data_tape.head = data_tape.head - 1 
            end 
        end,
        ['+'] = function()
            data_tape.tape[data_tape.head] = (data_tape.tape[data_tape.head] + 1) % 256
        end,
        ['-'] = function()
            if data_tape.tape[data_tape.head] > 0 then
                data_tape.tape[data_tape.head] = data_tape.tape[data_tape.head] - 1 
            else 
                data_tape.tape[data_tape.head] = 255 
            end 
        end,
        [','] = function() 
            data_tape.tape[data_tape.head] = string.byte(io.read("*line"))
        end,
        ['.'] = function()
            io.write(string.char(data_tape.tape[data_tape.head]))
        end,
        ['['] = function(head)
            if data_tape.tape[data_tape.head] == 0 then
                local depth = 0
                while head <= #src do
                    if src:sub(head, head) == '[' then
                        depth = depth + 1
                    elseif src:sub(head, head) == ']' then
                        depth = depth - 1
                    end

                    if depth == 0 then
                        break
                    end

                    head = head + 1
                end

                return head
            end
        end,
        [']'] = function(head)
            if data_tape.tape[data_tape.head] ~= 0 then
                local depth = 0
                while head >= 0 do
                    if src:sub(head, head) == '[' then
                        depth = depth + 1
                    elseif src:sub(head, head) == ']' then
                        depth = depth - 1
                    end

                    if depth == 0 then
                        break
                    end

                    head = head - 1
                end

                return head
            end
        end
    }

    local head = 0

    while head >= 0 and head <= #src do
        local handler = instructions[src:sub(head, head)]
        if handler then
            local temp = handler(head)
            if temp then
                head = temp
            end
        end

        head = head + 1
    end

end

function repl()
    print("BrainFuck Interpreter\n\nType 'bye' to exit.")
    while true do
        io.write("$ ")
        local line = io.read("*line")

        if line == "bye" then
            break
        end

        interpreter(line)
    end

    os.exit(true, true)
end

main()

File: day7/bf.js
import fs from 'fs';

let tape_head = 0
let tape_size = 256
let tape = []

function fgets(maxlen = 1024){
    const buf = Buffer.alloc(1);
    let result = '';

    for(let i = 0; i < maxlen - 1; i++){
        let n;
        try{
            n = fs.readSync(0, buf, 0, 1);
        }
        catch{
            break;
        }
        if(n === 0)
            break;  // EOF
        result += buf.toString();
        if(buf[0] === 0x0a)
            break;  // '\n'
    }

    return result.length ? result : null;  // null on EOF, like C
}

function fread(path) {
    try { return fs.readFileSync(path, 'utf8'); }
    catch { return null; }
}

function print(str){
    process.stdout.write(str)
}

function check_brace_balance(src){
    let balance = 0;

    let head = 0;
    while(head < src.length){
        if(src[head] == '[')
            balance++
        else if(src[head] == ']')
            balance--

        if(balance < 0)
            break

        head += 1
    }

    return balance
}

function interpreter(src){
    let balance = check_brace_balance(src)
    if(balance < 0){
        print("[ERROR]: Unbalanced ']'.")
        process.exit(1)
    }
    else if(balance > 0){
        print("[ERROR]: Unbalanced '['.")
        process.exit(1)
    }

    let head = 0;
    while(head < src.length){
        switch(src[head]){
            case '>':{
                tape_head++;
                if(tape_head >= tape_size){
                    tape_size *= 2
                    for(let i = tape_size / 2; i < tape_size; i++)
                        tape.push(0)
                }
                break;
            }
            case '<':{
                if(tape_head == 0)
                    break
                tape_head--
                break;
            }
            case '+':{
                tape[tape_head] = (tape[tape_head] + 1) % 256
                break
            }
            case '-':{
                if(tape[tape_head] == 0){
                    tape[tape_head] = 255
                    break
                }

                tape[tape_head]--
                break
            }
            case '.':{
                print(String.fromCharCode(tape[tape_head]))
                break
            }
            case ',':{
                tape[tape_head] = fgets().charCodeAt(0)
                break
            }
            case '[':{
                if(tape[tape_head] == 0){
                    let depth = 0
                    while(head < src.length){
                        if(src[head] == '[')
                            depth++
                        else if(src[head] == ']')
                            depth--

                        if(depth == 0)
                            break

                        head++
                    }
                }
                break
            }
            case ']':{
                if(tape[tape_head] != 0){
                    let depth = 0
                    while(head >= 0){
                        if(src[head] == '[')
                            depth++
                        else if(src[head] == ']')
                            depth--

                        if(depth == 0)
                            break

                        head--
                    }
                }
                break
            }
        }
        
        head++;
    }
}

function main(){
    const args = process.argv
    if(args.length > 1){
        let src = fread(args[2])

        // print(src)

        interpreter(src)

        return
    }

    // init the tape
    for(let i = 0; i < tape_size; i++)
        tape.push(0)

    // repl
    print("BrainFuck Interpreter\n\nType 'bye' to exit.\n");
    while(true){
        print("$ ")
        let line = fgets();


        interpreter(line)
    }
}

main()